text/0000-post-installation-mirror.md

```diff
@@ -0,0 +1,124 @@
+- Start Date: 2017-02-16
+- RFC PR:
+- Yarn Issue:
+
+# Summary
+
+Enable storing for post installation artifacts with Yarn offline mirror.
+
+# Motivation
+
+`Yarn` has improved node dependency resolution reliability dramatically.  However, we still can not
+achieve truly deterministic dependency resolution and hermetic build without addressing issues with installation scripts.
+
+Node modules allow arbitrary code execution before, during and after installation.  Although this feature has provided enormous flexibility for package owners, it is also one of the root causes for non-deterministic build.  Some packages use installation scripts to download additional codes and / or dependencies. This means those packages cannot be installed in network restricted environments (NRE). Production zones in enterprise environments are one of the examples of those NREs. They tend to restrict network access for obvious security reasons.
+
+Without further discussion with validity of installation scripts, a possible workaround can be achieved by storing the post installation artifacts using yarn offline mirror. In essence, we are aiming to provide a better alternative than an already prevalent practice among enterprise users, namely, save or check in the post installation node_modules folder.  Formalize this practice provides at least following benefits:
+- Allows node build to be truly offline, deterministic and hermetic
+- Allows package owners to continue use installation scripts
+- Allows saving of a single .tar.gz file instead of hundreds, if not thousands of files
+- Easier code review for addition, update and deletion of dependencies
+
+# Detailed design
+
+## Assumptions
+- Most, if not all, installation scripts only modify files and directories within its own folders.
+- Storing post installation artifacts is voluntary on per package basis.
+
+## Analysis of installation scripts
+It is difficult to categorized all possible usages of installation scripts as they are truly free formed. That being said, there are a couple of common patterns have been observed so far:
+- Download additional dependencies or codes
+- Compile node native extensions
+
+This observation leads to the assumption that most installation scripts only modify files and directories within its own folders.  Consequently, we can store the post installation content as artifacts without worrying about inter-module dependencies.
+
+##Modification to `yarn` offline mirror structure and `yarn.lock`

```

Comment:
Offline mirror kicks in at `fetch` phase. 
After the .tgz file is extracted into global cache folder `link` phase starts.
During `link` phase files are copied from cache into node_modules, considering hoisting, and then `lifecycle scripts` are executed that modify some files on those node_modules.

You would have to generate a new .tgz file for each package folder that got modified after `lifecycle scripts` phase, disabling their `lifecycle scripts`, and then modify yarn.lock file to point to the new .tgz file.

That could be quite complex to implement without bringing too much complexity into Yarn.